### ISSUE TITLE
尝试新的匹配方式

### ISSUE THREAD

**Issue by Melchoirr** [Melchoirr/xiaoyi-query#11]
1. 加入匹配手段：时间等
2. 加入pred分布先验

**Comment by Melchoirr** [comment]
## 检索匹配实验更新

对应提交：`dd56151`，提交信息为 `Add retrieval experiments and documentation`。

### 方法概述

当前主实验是一个 ETT 上的 channel-independent full-bank retrieval forecaster。对每个 test window、每个 channel，分别在完整 train bank 中检索 top-k 相似历史窗口，然后取这些候选未来窗口做加权平均预测。

baseline 只使用 shape loss。rerank 使用下面的组合分数：

```text
shape_weight * row_norm(shape_loss)
  + direction_weight * row_norm(same_direction_loss)
  + raw_weight * row_norm(raw_loss)
  + std_weight * row_norm(std_loss)
```

核心设计：

- `shape_loss`：比较历史窗口的归一化形状。
- `same_direction_loss`：比较一阶差分方向。query 和 candidate 在更多可比较时间步上同向变化时，loss 更小。
- `row_norm`：对每个 query、每个 channel 的候选 loss 做 row-median normalization，让不同 loss 的尺度可比较。
- 当前主要参数：`seq_len=96`，`top_k=50`，`direction_weight=1.5`。
- 预测聚合：对选中的 top-k future 做 inverse-distance weighted average。
- 聚合前默认使用 `--align_mode clipped_std`，按历史窗口统计量对齐 candidate future。
- torch 后端支持 `cpu/cuda/mps`；`--channel_backend batched` 会把 7 个 channel 放在同一个 torch tensor 中并行计算，但 top-k 仍然按 channel 独立选择，不会混在一起。

### 关键代码位置和结构

- `experiments/README.md`
  - 方法说明、运行命令、复现实验说明。
- `experiments/data_loader.py`
  - 共享数据加载文件，风格对齐 Time-Series-Library。
  - 支持 `ETTh1`、`ETTh2`、`ETTm1`、`ETTm2` 和 `custom`。
  - 提供 `data_provider`、`collect_windows`、`build_shape_features`。
- `experiments/future_factor_retrieval_etth1/channel_independent_direction_retrieval_standalone.py`
  - 当前主实验文件。
  - 包含检索打分、same-direction loss、row norm、top-k、future alignment、metric 汇总。
  - 重要参数：`--weights`、`--top_k`、`--backend`、`--channel_backend`、`--selected_batch_size`。
- `scripts/experiments/run_mps_ett_four_datasets_four_lengths.sh`
  - 一键跑四个 ETT 数据集 x 四个预测长度。
- `scripts/experiments/run_mps_vs_numpy_etth1.sh`
  - ETTh1 上的 MPS/numpy 对照 smoke check。
- `outputs/experiment_audit/seq96_best_results.csv`
  - 当前整理后的最佳结果表。
- `outputs/mps_ett_four_datasets_four_lengths/*/summary.csv`
  - 当前 MPS sweep 的各组 summary。

PatchTST 相关实验保留在：

- `experiments/patchtst_ssl_embedding_retrieval_etth1/`

这部分只复用 `experiments.data_loader`，不依赖 standalone matching 实验目录。

### 运行方式

单组实验示例：

```bash
/Users/zhaodawei/miniconda3/envs/agent/bin/python \
  experiments/future_factor_retrieval_etth1/channel_independent_direction_retrieval_standalone.py \
  --root_path dataset \
  --data_path ETTh1.csv \
  --features M \
  --target OT \
  --seq_len 96 \
  --label_len 48 \
  --pred_len 96 \
  --flag test \
  --weights 1.5 \
  --top_k 50 \
  --backend mps \
  --channel_backend batched \
  --selected_batch_size 32 \
  --output_dir outputs/mps_ett_four_datasets_four_lengths/ETTh1_sl96_pl96
```

完整 sweep：

```bash
bash scripts/experiments/run_mps_ett_four_datasets_four_lengths.sh
```

如果 MPS 内存不够，可以降低 query batch size：

```bash
SELECTED_BATCH_SIZE=16 bash scripts/experiments/run_mps_ett_four_datasets_four_lengths.sh
```

### 最佳结果汇总

结果来源：`outputs/experiment_audit/seq96_best_results.csv`。

| dataset | pred_len | method | MSE | MAE | 权重 |
|---|---:|---|---:|---:|---|
| ETTh1 | 96 | rerank_ci | 0.366692 | 0.394583 | direction=1.5 |
| ETTh1 | 192 | rerank_ci | 0.413981 | 0.422466 | direction=1.5 |
| ETTh1 | 336 | shape_baseline_ci | 0.454928 | 0.448161 | baseline |
| ETTh1 | 720 | shape_baseline_ci | 0.466913 | 0.471760 | baseline |
| ETTh2 | 96 | rerank_ci | 0.292957 | 0.349835 | direction=1.5 |
| ETTh2 | 192 | rerank_ci | 0.369660 | 0.397676 | direction=1.5, raw=0.1, std=0.1 |
| ETTh2 | 336 | rerank_ci | 0.408950 | 0.423267 | direction=1.5, raw=0.1, std=0.1 |
| ETTh2 | 720 | rerank_ci | 0.436506 | 0.446018 | direction=1.5, raw=0.1, std=0.1 |
| ETTm1 | 96 | rerank_ci | 0.301227 | 0.351909 | direction=1.5 |
| ETTm1 | 192 | rerank_ci | 0.344887 | 0.373508 | direction=1.5 |
| ETTm1 | 336 | rerank_ci | 0.380487 | 0.396990 | direction=1.5 |
| ETTm1 | 720 | rerank_ci | 0.440198 | 0.430869 | direction=1.5 |
| ETTm2 | 96 | rerank_ci | 0.196018 | 0.278465 | direction=1.5, raw=0.1, std=0.1 |
| ETTm2 | 192 | rerank_ci | 0.248783 | 0.310107 | direction=1.5, raw=0.1, std=0.1 |
| ETTm2 | 336 | rerank_ci | 0.302992 | 0.346196 | direction=1.5, raw=0.1, std=0.1 |
| ETTm2 | 720 | rerank_ci | 0.746805 | 0.477709 | direction=1.5, raw=0.1, std=0.1 |

### 备注

- ETT 切分逻辑在 `experiments/data_loader.py` 中按 Time-Series-Library 的 border 设置实现。
- `label_len` 主要是为了兼容 dataloader 形式；当前 retrieval 预测实际使用的是收集到的 history/future windows。
- torch batched 路径已经和 loop 路径做过小规模对照：baseline 完全一致，rerank 只存在约 `1e-6` 量级差异，主要来自 float/top-k 细节。
- 图片输出已加入 `.gitignore`，仓库内主要保留代码、脚本和 CSV 结果用于复现。
